internal/bus/kafka.go

```diff
@@ -0,0 +1,204 @@
+			Username: username,
+			Password: password,
+		}, nil
+	case "scram-sha-512":
+		mechanism, err := scram.Mechanism(scram.SHA512, username, password)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create scram-sha-512 mechanism: %w", err)
+		}
+
+		return mechanism, nil
+	case "scram-sha-256":
+		mechanism, err := scram.Mechanism(scram.SHA256, username, password)
+		if err != nil {
+			return nil, fmt.Errorf("unable to create scram-sha-256 mechanism: %w", err)
+		}
+
+		return mechanism, nil
+	default:
+		return nil, fmt.Errorf("%w: %s", UnknownSaslMechanismErr, saslMechanismName)
+	}
+}
+
+func InitializeKafkaBroker() error {
+	var err error
+	broker, err = NewKafkaBroker()
+	if err != nil {
+		return fmt.Errorf("unable to initialize kafka: %w", err)
+	}
+
+	return nil
+}
+
+func NewKafkaBroker() (Broker, error) {
+	var tlsConfig *tls.Config
+	var saslMechanism sasl.Mechanism
+
+	// configure TLS when CA certificate was provided
+	if config.Kafka.CACert != "" {
+		cleanCAPath := filepath.Clean(config.Kafka.CACert)
+		pemCerts, err := ioutil.ReadFile(cleanCAPath)
+		if err != nil {
+			return nil, fmt.Errorf("unable to read kafka CA cert: %w", err)
+		}
+
+		pool := x509.NewCertPool()
+		pool.AppendCertsFromPEM(pemCerts)
+
+		tlsConfig = &tls.Config{
+			MinVersion: tls.VersionTLS13,
+			RootCAs:    pool,
+		}
+	}
+
+	// configure SASL if mechanism was provided
+	if config.Kafka.SASL.SaslMechanism != "" {
+		var err error
+		saslMechanism, err = createSaslMechanism(config.Kafka.SASL.SaslMechanism, config.Kafka.SASL.Username, config.Kafka.SASL.Password)
+		if err != nil {
+			return nil, fmt.Errorf("kafka SASL error: %w", err)
+		}
+	}
+
+	dialer := &kafka.Dialer{
+		ClientID:      version.KafkaClientID,
+		Timeout:       10 * time.Second,
+		SASLMechanism: saslMechanism,
+		TLS:           tlsConfig,
+	}
+
+	transport := &kafka.Transport{
+		Dial:     dialer.DialFunc,
+		ClientID: version.KafkaClientID,
+		TLS:      tlsConfig,
+		SASL:     saslMechanism,
+	}
+
+	return &kafkaBroker{
+		dialer:    dialer,
+		transport: transport,
+	}, nil
+}
+
+func newContextLogger(ctx context.Context) func(msg string, a ...interface{}) {
+	return func(msg string, a ...interface{}) {
+		logger := ctxval.Logger(ctx)
+		logger.Debug().Bool("kafka", true).Msgf(msg, a...)
+	}
+}
+
+func newContextErrLogger(ctx context.Context) func(msg string, a ...interface{}) {
+	return func(msg string, a ...interface{}) {
+		logger := ctxval.Logger(ctx)
+		logger.Warn().Bool("kafka", true).Msgf(msg, a...)
+	}
+}
+
+// NewReader creates a reader. Use Close() function to close the reader.
+func (b *kafkaBroker) NewReader(ctx context.Context, topic string) *kafka.Reader {
+	return kafka.NewReader(kafka.ReaderConfig{
+		Brokers:     config.Kafka.Brokers,
+		Dialer:      b.dialer,
+		Topic:       topic,
+		StartOffset: kafka.LastOffset,
+		Logger:      kafka.LoggerFunc(newContextLogger(ctx)),
+		ErrorLogger: kafka.LoggerFunc(newContextErrLogger(ctx)),
+	})
+}
```

Comment:
One thing tho - when statuser is started, it should start reading from the end (last offset), we do not want to read the whole history of the topic. However, during my testing it looks like when I start a new reader, it reads the whole history regardless of the `StartOffset` setting. Do you see the same behavior? I am not sure why I see that.

Reply:
When testing this, keep this in mind - then there are existing messages in a topic and a worker starts, it should NOT process them, only new ones. It does not work this way currently, I don't know why yet.